querybook/webapp/lib/python/usePython.ts

```diff
@@ -9,44 +11,78 @@ import {
 
 interface UsePythonProps {
     docId?: number;
+    cellId?: number;
     onStdout?: (message: string) => void;
     onStderr?: (message: string) => void;
     onComplete?: () => void;
 }
 interface UsePythonReturn {
     kernelStatus: PythonKernelStatus;
     runPython: (code: string) => Promise<void>;
-    cancelRun: () => void;
+    cancelRun: () => Promise<void>;
     createDataFrame: (
         dfName: string,
         statementExecutionId: number,
         namespaceId?: number
     ) => Promise<void>;
     stdout: string[];
-    stderr: string[];
+    stderr: string | null;
     executionStatus: PythonExecutionStatus;
     executionCount: number;
     getNamespaceInfo: (namespaceId: number) => Promise<PythonNamespaceInfo>;
 }
 
 export default function usePython({
     docId,
+    cellId,
     onStdout,
     onStderr,
 }: UsePythonProps): UsePythonReturn {
     const [stdout, setStdout] = useState<string[]>([]);
-    const [stderr, setStderr] = useState<string[]>([]);
+    const [stderr, setStderr] = useState<string>();
     const [executionStatus, setExecutionStatus] =
         useState<PythonExecutionStatus>();
     const [executionCount, setExecutionCount] = useState<number>();
+    const cancelPromiseResolveRef = useRef(null);
 
     const { status, runPython, cancelRun, createDataFrame, getNamespaceInfo } =
         useContext(PythonContext);
 
+    useEffect(() => {
+        if (cellId) {
+            PythonCellResource.getResult(cellId).then((resp) => {
+                const result = resp.data;
+                if (result) {
+                    setStdout(result.output);
+                    setStderr(result.error);
+                }
+            });
+        }
+    }, []);
+
+    useEffect(() => {
+        if (
+            cellId &&
+            (executionStatus === PythonExecutionStatus.SUCCESS ||
+                executionStatus === PythonExecutionStatus.ERROR ||
+                executionStatus === PythonExecutionStatus.CANCEL)
+        ) {
+            PythonCellResource.updateResult(cellId, stdout, stderr);
+        }
+    }, [executionStatus, stdout, stderr]);
```

Comment:
this useEffect already triggers on stdout/stderr changes, are we basically make sure no intermediate states are saved?

Reply:
to only trigger the result save to db only happens once (after the cell execution finishes). stdout may change multiple times for the cell run.

Reply:
hmm, yeah I was wondering if we should just do a debounced save instead